src/main/java/com/exasol/bucketfs/uploadnecassity/ChecksumUploadNecessityCheckStrategy.java

```diff
@@ -0,0 +1,144 @@
+
+/**
+ * This {@link UploadNecessityCheckStrategy} compares the checksum of the local file and the file on BucketFs and only
+ * uploads the file if they differ or if the file does not exist in BucketFs.
+ */
+//[impl->dsn~conditional-upload~1]
+public class ChecksumUploadNecessityCheckStrategy implements UploadNecessityCheckStrategy {
+    private static final String UDF_SCHEMA = "BUCKET_FS_JAVA_HELPER";
+    private static final String UDF_NAME = "bucketfs_md5_sum";
+    private static final String UDF_FULL_NAME = UDF_SCHEMA + "." + UDF_NAME;
+    private static final int ONE_MEGABYTE = 1000000;
+    private final Connection sqlConnection;
+
+    /**
+     * Create a new instance of {@link ChecksumUploadNecessityCheckStrategy}.
+     * 
+     * @param sqlConnection SQL connection to the Exasol database. (used for calculating the checksum of the file in
+     *                      BucketFs in a UDF)
+     */
+    public ChecksumUploadNecessityCheckStrategy(final Connection sqlConnection) {
+        this.sqlConnection = sqlConnection;
+    }
+
+    @Override
+    public boolean isUploadNecessary(final Path file, final String fullFileNameInBucketFs, final ReadOnlyBucket bucket)
+            throws BucketAccessException {
+        try {
+            final String[] parts = fullFileNameInBucketFs.split(BucketConstants.PATH_SEPARATOR);
+            final String fileName = parts[parts.length - 1];
+            final List<String> filesInBucketDirectory = bucket.listContents(getDirectory(parts));
+            if (Files.size(file) > ONE_MEGABYTE && filesInBucketDirectory.contains(fileName)) {
+                return !localMd5Checksum(file).equals(getMd5Checksum(fullFileNameInBucketFs, bucket));
+            } else {
+                return true;
+            }
+        } catch (final BucketAccessException | NoSuchAlgorithmException | IOException exception) {
+            throw new BucketAccessException(ExaError.messageBuilder("E-BFSJ-17")
+                    .message("Failed to check if we need to upload {{file}}.", fullFileNameInBucketFs).toString(),
+                    exception);
+        }
+    }
+
+    private String getDirectory(final String[] parts) {
+        final String directory = Arrays.stream(parts).limit(parts.length - 1)
+                .collect(Collectors.joining(BucketConstants.PATH_SEPARATOR));
+        if (directory.startsWith(BucketConstants.PATH_SEPARATOR)) {
+            return directory.substring(1);
+        } else {
+            return directory;
+        }
+    }
+
+    private String localMd5Checksum(final Path localPath) throws NoSuchAlgorithmException, IOException {
+        final MessageDigest checksumBuilder = MessageDigest.getInstance("MD5");
+        try (final InputStream inputStream = Files.newInputStream(localPath);
+                final DigestInputStream checksumBuildingStream = new DigestInputStream(inputStream, checksumBuilder)) {
+            final byte[] buffer = new byte[1000];
```

Comment:
Why a fixed size of 1000? Why not 999 or 1001?

Reply:
I'm a human. I like round numbers even though it's not rational.